src/cc/libbpf.c

```diff
@@ -400,43 +426,78 @@ int bpf_prog_load(enum bpf_prog_type prog_type, const char *name,
     }
   }
 
-  if (ret < 0 && !log_buf) {
+  if (ret < 0) {
+    if (log_buf) {
+      // If user alraedy provided log buffer, just print message and return.
+      bpf_print_hints(ret, log_buf);
+      if (errno == ENOSPC)
+        fprintf(stderr, "bpf: log_buf size may be insufficient\n");
+      goto return_result;
+    }
 
-    buffer_size = LOG_BUF_SIZE;
-    // caller did not specify log_buf but failure should be printed,
-    // so repeat the syscall and print the result to stderr
+    // User did not provide log buffer. We will try to increase size of
```

Comment:
Do you want to check whether the error is caused by insufficient buffer or not here?

Reply:
At this point, we already know the load syscall failed, and user did not provide a buffer. Regardless of the reason, we will try allocate temporary buffer size from `LOG_BUF_SIZE`, and increase it until `ENOSPC` no longer happens.

At the worse case is user have specified `log_level = 1` and we already tried temporary buffer size being `LOG_BUF_SIZE` and it turned out not to be enough. In such case, we just do an extra iteration of attempted load. It only happens on error and if we want to get around that this part of logic would be more complicated. I don't feel it's worth it......